go/pkg/pwagent/daemon.go

```diff
@@ -56,28 +64,121 @@ func Daemon(ctx context.Context, cli *client.Client, logger *zap.Logger) error {
+							found = true
+							if apiInstance.GetStatus() == pwdb.ChallengeInstance_NeedRedump {
+								needRedump = true
+							}
+						}
+					}
+					seasonChallenges := apiInstanceFlavor.GetSeasonChallenges()
+					if seasonChallenges != nil {
+						for _, seasonChallenge := range seasonChallenges {
+							subscriptions := seasonChallenge.GetSubscriptions()
+							if subscriptions != nil {
+								for _, subscription := range subscriptions {
+									if subscription.GetStatus() == pwdb.ChallengeSubscription_Active {
+										team := subscription.GetTeam()
+										if team != nil {
+											members := team.GetMembers()
+											if members != nil {
+												for _, member := range members {
+													for _, instance := range flavor.Instances {
+														for _, port := range instance.Ports {
+															if port.PublicPort != 0 {
+																// configure nginx
+																// - compute the list of members in Active teams
+																// - generate a hash based on their userIDs
+																instanceName := instance.Names[0][1:]
+																_, entryFound := agentOpts.AllowedUsers[instanceName]
+																if !entryFound {
+																	agentOpts.AllowedUsers[instanceName] = []int64{member.GetID()}
+																} else {
+																	allowedUsersSlice := agentOpts.AllowedUsers[instanceName]
+																	allowedUsersSlice = append(allowedUsersSlice, member.GetID())
+																	agentOpts.AllowedUsers[instanceName] = allowedUsersSlice
+																}
+															}
+														}
+													}
+												}
+											}
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
 
-	// start missing instances
+			// start missing instances
+			// redump instances with state=NeedRedump
+			if !found || needRedump {
+				composeFile, err := pwcompose.Prepare(
+					path.Join(defaultChallengesPath, apiInstance.GetFlavor().GetChallenge().GetName()),
+					"pathwar/",
+					false,
+					apiInstance.GetFlavor().GetVersion(),
+					logger,
+				)
+				if err != nil {
+					return errcode.ErrNotImplemented.Wrap(err)
+				}
 
-	// redump instances with state=NeedRedump
+				var configData pwinit.InitConfig
+				err = json.Unmarshal(apiInstance.InstanceConfig, &configData)
+				if err != nil {
+					return errcode.ErrNotImplemented.Wrap(err)
+				}
 
-	// configure nginx
-	// - compute the list of members in Active teams
-	// - generate a hash based on their userIDs
+				err = pwcompose.Up(ctx, composeFile, "default", needRedump, &configData, cli, logger)
+				if err != nil {
+					return errcode.ErrNotImplemented.Wrap(err)
+				}
+			}
+		}
 
-	// later: for each updated instances -> call api to update status
+		err = Nginx(ctx, agentOpts, cli, logger)
+		if err != nil {
+			return errcode.ErrNotImplemented.Wrap(err)
+		}
 
-	// last step -> make this a daemon: -> loop every X seconds
+		if runOnce {
```

Comment:
I suggest you to create a dedicated method that will do stuff "once" and then make this entrypoint function react differently if running with runOnce or without

RunOnce will exit, so it's a nice thing to return an error

Without RunOnce should not quit (except if something very important happens, but we can see this later), so for now you can just log an error but don't kill the daemon